maven_iuvs/echelle.py

```diff
@@ -588,26 +753,50 @@ def subtract_darks(light_fits, dark_fits):
 
     light_data = light_fits['Primary'].data
 
-    # Retrieve dark frames
+    # Retrieve dark frames. Note that get_dark_frames checks to see if the dark
+    # file only has one frame. If that's the case, it creates a second frame, 
+    # but sets it to nan. 
     darks = get_dark_frames(dark_fits)
     first_dark = darks[0, :, :]
     second_dark = darks[1, :, :]
 
-    # check darks are valid
-    if np.isnan(first_dark).any() & np.isnan(second_dark).any():
-        raise Exception(f"Missing critical observation data: no valid darks")
-
     # Make the array to store dark-subtracted data
     dark_subtracted = np.zeros_like(light_data)
-    
-    # Get rid of extra frames where light data are bad (broken or nan)
+
+    # Set up tracking arrays
     medians = []
     good_frame_inds = [] # Frames which appear to have valid data
     nan_light_inds = [] # frames with NaN
     bad_light_inds = [] # Light frames which have some problem, i.e. oversaturation, but are not NaN
     nan_dark_inds = []  # Dark frames with NaN 
     light_frames_with_nan_dark = [] # Light frames whose dark frame is NaN
 
+    # CHECK VALIDITY OF DARKS 
+    # =========================================================================
+    first_dark_good = True
+    second_dark_good = True
+
+    # # Check goodness of first dark 
+    if (np.isnan(first_dark).any()) or (np.nanmedian(first_dark) > 1000):  # First dark is bad 
+        first_dark_good = False
+    
+    # Check that second dark exists - early in mission, we didn't always take 2 darks.
+    # In those cases we should use the first dark.
+    second_dark_exists = True 
+    if np.isnan(second_dark).all():
```

Comment:
add a comment: `# an all nan second_dark is guaranteed by get_dark_frames() above if only one dark frame exists`

Reply:
added in 84cb3f619af0f201db32140ec823d57714f2519c